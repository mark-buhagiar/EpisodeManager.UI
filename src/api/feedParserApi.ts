import axios from 'axios';
import ApiConfig from '../config/apiConfig';
import ParseHistoryListing from '../models/ParseHistoryListing';
import ParseHistoryDetails from '../models/ParseHistoryDetails';
import moment from 'moment';

const buildEndpoint = (endpoint: string): string => ApiConfig.feedParserApiBase + endpoint;

const feedParserMSEndpoints = {
    syncWithShowRss: buildEndpoint('sync/SyncWithShowRss'),
    forceSyncParseHistory: buildEndpoint('sync/ForceSyncParseHistory'),
    getParseHistoryBetweenDates: buildEndpoint('ParseHistory/GetParseHistoryBetweenDates'),
    getParseHistoryDetails: buildEndpoint('ParseHistory/GetParseHistoryDetails'),
};

export async function syncWithShowRss(): Promise<void> {
    try {
        await axios.post(feedParserMSEndpoints.syncWithShowRss);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

function convertDates(listing: ParseHistoryListing): ParseHistoryListing {
    return {
        ...listing,
        retrievedOn: moment.utc(listing.retrievedOn).toDate(),
        lastSyncedOn: moment.utc(listing.lastSyncedOn).toDate(),
    };
}

export async function getParseHistoryBetweenDates(startDate: Date, endDate: Date): Promise<ParseHistoryListing[]> {
    try {
        const params = {
            startDate,
            endDate,
        };
        const parseHistory = await axios.get<ParseHistoryListing[]>(feedParserMSEndpoints.getParseHistoryBetweenDates, {
            params,
        });

        return parseHistory.data.map(convertDates);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getParseHistoryDetails(id: number): Promise<ParseHistoryDetails> {
    try {
        const params = {
            parseHistoryId: id,
        };
        const parseHistoryDetails = await axios.get<ParseHistoryDetails>(feedParserMSEndpoints.getParseHistoryDetails, {
            params,
        });

        return parseHistoryDetails.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
export async function forceSyncParseHistory(id: number): Promise<void> {
    try {
        const params = {
            parseHistoryId: id,
        };

        await axios.post(feedParserMSEndpoints.forceSyncParseHistory, null, {
            params,
        });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
