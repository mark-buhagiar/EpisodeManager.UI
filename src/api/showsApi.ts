import axios from 'axios';
import Episode from '../models/Episode';
import ApiConfig from '../config/apiConfig';

const buildEndpoint = (endpoint: string): string => ApiConfig.showsApiBase + endpoint;

const showsMSEndpoints = {
    getForCurrentUserBetweenDates: buildEndpoint('episodes/getForCurrentUserBetweenDates'),
};

export async function getForCurrentUserBetweenDates(startDate: Date, endDate: Date): Promise<Episode[]> {
    try {
        const params = {
            startDate: startDate,
            endDate: endDate,
        };
        const result = await axios.get<Episode[]>(showsMSEndpoints.getForCurrentUserBetweenDates, { params });
        return result.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
