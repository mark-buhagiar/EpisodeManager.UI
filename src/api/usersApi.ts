import axios from 'axios';
import ApiConfig from '../config/apiConfig';
import EpisodePreferences from '../models/EpisodePreferences';
import Show from '../models/Show';

const buildEndpoint = (endpoint: string): string => ApiConfig.usersApiBase + endpoint;

const usersMSEndpoints = {
    addEpisodeDownloaded: buildEndpoint('subscriptions/AddEpisodeDownloaded'),
    getUserSubscriptionsForListing: buildEndpoint('subscriptions/GetUserSubscriptionsForListing'),
    unsubscribeFromShow: buildEndpoint('subscriptions/UnsubscribeFromShow'),
    subscribeToShow: buildEndpoint('subscriptions/SubscribeToShow'),
    setUserEpisodePreferences: buildEndpoint('users/setUserEpisodePreferences'),
    getUserEpisodePreferences: buildEndpoint('users/getUserEpisodePreferences'),
};

export async function addEpisodeDownloaded(episodeId: number): Promise<void> {
    try {
        const params = {
            episodeId,
        };
        await axios.put(usersMSEndpoints.addEpisodeDownloaded, null, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getUserSubscriptionsForListing(): Promise<Show[]> {
    try {
        const result = await axios.get(usersMSEndpoints.getUserSubscriptionsForListing);
        return result.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function unsubscribeFromShow(showId: number): Promise<void> {
    try {
        const params = {
            showId,
        };
        await axios.patch(usersMSEndpoints.unsubscribeFromShow, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function subscribeToShow(showIds: number[]): Promise<void> {
    try {
        const params = {
            showIds,
        };
        await axios.put(usersMSEndpoints.subscribeToShow, null, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function setUserEpisodePreferences(preferences: EpisodePreferences): Promise<void> {
    try {
        await axios.post(usersMSEndpoints.setUserEpisodePreferences, preferences);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getUserEpisodePreferences(): Promise<EpisodePreferences> {
    try {
        const response = await axios.get(usersMSEndpoints.getUserEpisodePreferences);
        return response.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
