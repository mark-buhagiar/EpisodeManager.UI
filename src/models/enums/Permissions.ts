enum Permissions {
    READ_ADMIN = 'read:admin',
    EXECUTE_ADMIN = 'execute:admin',
}

export default Permissions;
