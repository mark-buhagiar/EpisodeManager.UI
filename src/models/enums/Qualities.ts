import Color from 'color';

export enum QualityId {
    StandardDef = 1,
    P720 = 2,
    P1080 = 3,
}

export enum QualityDescription {
    StandardDef = 'Standard definition',
    P720 = '720p',
    P1080 = '1080p',
}

export type Quality = {
    id: QualityId;
    description: QualityDescription;
};

export const QualityColors = {
    [QualityId.StandardDef]: Color('rgb(255, 99, 132)'),
    [QualityId.P720]: Color('rgb(255, 206, 86)'),
    [QualityId.P1080]: Color('rgb(75, 192, 192)'),
};

export const Qualities = [
    { id: QualityId.StandardDef, description: QualityDescription.StandardDef },
    { id: QualityId.P720, description: QualityDescription.P720 },
    { id: QualityId.P1080, description: QualityDescription.P1080 },
];
