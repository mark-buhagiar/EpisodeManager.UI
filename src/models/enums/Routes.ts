export enum Routes {
    HOMEPAGE = '/',
    MANAGE_SHOWS = '/manage',
    SHOWS = '/shows',
    HELP = '/help',
    ADMIN = '/admin',
    PROFILE = '/profile',
}

export enum RouteLabels {
    HOMEPAGE = 'Calendar',
    MANAGE_SHOWS = 'Manage Shows',
    SHOWS = 'Shows',
    HELP = 'Help',
    ADMIN = 'Admin',
    PROFILE = 'Profile',
}
