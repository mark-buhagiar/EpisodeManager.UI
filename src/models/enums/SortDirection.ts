enum SortDirection {
    ASCENDING = 1,
    DESCENDING = 2,
}

export default SortDirection;
