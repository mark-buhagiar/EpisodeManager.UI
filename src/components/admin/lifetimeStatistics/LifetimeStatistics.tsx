import React, { useState, useEffect } from 'react';
import PromiseStatus from './../../../models/PromiseStatus';
import { Qualities } from '../../../models/enums/Qualities';
import * as showsApi from '../../../api/showsApi';
import * as usersApi from '../../../api/usersApi';
import ComponentLoading from '../../loading/ComponentLoading';
import EpisodeCountQualityDistribution from '../../../models/EpisodeCountQualityDistribution';
import Panel from '../../common/panel';

const LifetimeStatistics: React.FC = (): JSX.Element => {
    const [isLoading, setIsLoading] = useState(false);
    const [userCount, setUserCount] = useState<number | null>(null);
    const [totalShowCount, setTotalShowCount] = useState<number | null>(null);
    const [episodeCount, setEpisodeCount] = useState<EpisodeCountQualityDistribution[] | null>(null);

    useEffect(() => {
        async function getData(): Promise<void> {
            setIsLoading(true);
            const usersPromise = usersApi.getTotalUserCount();
            const showsPromise = showsApi.getTotalShowCount();
            const episodesPromise = showsApi.getEpisodeCountQualityDistribution();

            const [users, shows, episodes] = await Promise.all(
                [usersPromise, showsPromise, episodesPromise].map((p: Promise<any>) =>
                    p
                        .then((result): PromiseStatus => ({ result, success: true } as PromiseStatus))
                        .catch((): PromiseStatus => ({ success: false } as PromiseStatus)),
                ),
            );

            setUserCount(users.success ? users.result : null);
            setTotalShowCount(shows.success ? shows.result : null);
            setEpisodeCount(episodes.success ? episodes.result : null);
            setIsLoading(false);
        }

        getData();
    }, []);

    return (
        <Panel title="Lifetime Statistics">
            {isLoading ? (
                <ComponentLoading />
            ) : (
                <>
                    <div>{`# of users: ${userCount ?? 'N/A'}`}</div>
                    <div>{`# of shows: ${totalShowCount ?? 'N/A'}`}</div>
                    {episodeCount?.map((qualityDistribution) => (
                        <div key={qualityDistribution.quality}>
                            {`# of episodes (${
                                Qualities.find((x) => x.id === qualityDistribution.quality)?.description ?? 'N/A'
                            }): ${qualityDistribution.count}`}
                        </div>
                    )) ?? 'Quality distribution N/A'}
                </>
            )}
        </Panel>
    );
};

export default LifetimeStatistics;
