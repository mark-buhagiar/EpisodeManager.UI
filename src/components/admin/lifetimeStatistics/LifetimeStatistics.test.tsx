import { cleanup, render, wait } from '@testing-library/react';
import React from 'react';
import { getEpisodeCountQualityDistribution, getTotalShowCount } from '../../../api/showsApi';
import { getTotalUserCount } from '../../../api/usersApi';
import { QualityId } from '../../../models/enums/Qualities';
import LifetimeStatistics from './LifetimeStatistics';

const totalUserCountResult = 12;
const totalShowCountResult = 11;
const qualityDistribution = [
    {
        count: 31,
        quality: 1,
    },
    {
        count: 31,
        quality: 2,
    },
    {
        count: 27,
        quality: 3,
    },
];

jest.mock('../../../api/usersApi', () => ({
    getTotalUserCount: jest.fn(),
}));

jest.mock('../../../api/showsApi', () => ({
    getEpisodeCountQualityDistribution: jest.fn(),
    getTotalShowCount: jest.fn(),
}));

afterEach(async () => {
    await cleanup();
    jest.useRealTimers();
    jest.clearAllMocks();
});

describe('When the component is rendered', () => {
    afterEach(async () => {
        await cleanup();
        jest.useRealTimers();
        jest.clearAllMocks();
    });

    it('should display a loading indicator while data is being retrieved', async (): Promise<void> => {
        (getEpisodeCountQualityDistribution as jest.Mock).mockResolvedValue(qualityDistribution);
        (getTotalUserCount as jest.Mock).mockResolvedValue(totalUserCountResult);
        (getTotalShowCount as jest.Mock).mockResolvedValue(totalShowCountResult);

        const { getByRole, asFragment } = render(<LifetimeStatistics />);

        expect(getByRole('progressbar')).toBeTruthy();
        expect(asFragment()).toMatchSnapshot();

        await wait();
    });

    it('should display a metadata once data is retrieved', async (): Promise<void> => {
        (getEpisodeCountQualityDistribution as jest.Mock).mockResolvedValue(qualityDistribution);
        (getTotalUserCount as jest.Mock).mockResolvedValue(totalUserCountResult);
        (getTotalShowCount as jest.Mock).mockResolvedValue(totalShowCountResult);

        const { asFragment, queryByRole } = render(<LifetimeStatistics />);
        await wait();

        expect(queryByRole('progressbar')).toBeNull();
        expect(asFragment()).toMatchSnapshot();
    });

    it.each`
        usersError | showsError | distributionError
        ${false}   | ${false}   | ${false}
        ${false}   | ${false}   | ${true}
        ${false}   | ${true}    | ${false}
        ${false}   | ${true}    | ${true}
        ${true}    | ${false}   | ${false}
        ${true}    | ${false}   | ${true}
        ${true}    | ${true}    | ${false}
        ${true}    | ${true}    | ${true}
    `(
        'should display successful requests if a request fails',
        async ({ usersError, showsError, distributionError }): Promise<void> => {
            const errorResult = new Error('test error');

            if (distributionError) (getEpisodeCountQualityDistribution as jest.Mock).mockRejectedValue(errorResult);
            else (getEpisodeCountQualityDistribution as jest.Mock).mockResolvedValue(qualityDistribution);

            if (showsError) (getTotalShowCount as jest.Mock).mockRejectedValue(errorResult);
            else (getTotalShowCount as jest.Mock).mockResolvedValue(totalShowCountResult);

            if (usersError) (getTotalUserCount as jest.Mock).mockRejectedValue(errorResult);
            else (getTotalUserCount as jest.Mock).mockResolvedValue(totalUserCountResult);

            const { getByText } = render(<LifetimeStatistics />);
            await wait();

            if (usersError) expect(getByText(`# of users: N/A`)).toBeTruthy();
            else expect(getByText(`# of users: ${totalUserCountResult}`)).toBeTruthy();

            if (distributionError) expect(getByText(`Quality distribution N/A`)).toBeTruthy();
            else {
                expect(
                    getByText(
                        `# of episodes (Standard definition): ${
                            qualityDistribution.find((x) => x.quality === QualityId.StandardDef)?.count
                        }`,
                    ),
                ).toBeTruthy();
                expect(
                    getByText(
                        `# of episodes (720p): ${qualityDistribution.find((x) => x.quality === QualityId.P720)?.count}`,
                    ),
                ).toBeTruthy();
                expect(
                    getByText(
                        `# of episodes (1080p): ${
                            qualityDistribution.find((x) => x.quality === QualityId.P1080)?.count
                        }`,
                    ),
                ).toBeTruthy();
            }

            if (showsError) expect(getByText(`# of shows: N/A`)).toBeTruthy();
            else expect(getByText(`# of shows: ${totalShowCountResult}`)).toBeTruthy();
        },
    );
});
