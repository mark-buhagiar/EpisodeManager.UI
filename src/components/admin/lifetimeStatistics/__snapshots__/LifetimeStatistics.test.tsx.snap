// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When the component is rendered should display a loading indicator while data is being retrieved 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Lifetime Statistics
      <div
        class="panel-actions"
      />
      <div
        class="separator"
      />
    </div>
    <div
      class="component-loading-container"
    >
      <span
        class="component-loading"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`When the component is rendered should display a metadata once data is retrieved 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Lifetime Statistics
      <div
        class="panel-actions"
      />
      <div
        class="separator"
      />
    </div>
    <div>
      # of users: 12
    </div>
    <div>
      # of shows: 11
    </div>
    <div>
      # of episodes (Standard definition): 31
    </div>
    <div>
      # of episodes (720p): 31
    </div>
    <div>
      # of episodes (1080p): 27
    </div>
  </div>
</DocumentFragment>
`;
